### ISSUE TITLE
Question: why does upsert replace record if no updates present?

### ISSUE THREAD

**Issue by felixSchl** [yesodweb/persistent#668]
I am referring to this implementation of upsert:

https://hackage.haskell.org/package/persistent-2.2.4/docs/src/Database-Persist-Class-PersistUnique.html#upsert

I am wondering what is the motivation for this special case: `when (null updates) (replace k record)`?
I was surprised that when my updates turned out to be empty, my record was completely reset.

Maybe I am using the function wrong, but here's what I am doing:

```
upsertUser :: (
      MonadIO m
    , BaseBackend backend ~ SqlBackend
    , PersistUniqueWrite backend
    )
    => Text
    -> Maybe (Maybe Text)
    -> ReaderT backend m (Entity User)
upsertUser email mmpassword = upsert fullRecord updates
  where
    fullRecord = User {
        userIdent = email
      , userPassword = Nothing
    }
    updates = catMaybes [ (UserPassword =.) <$> mmpassword ]
```

I think the whole issue arises from allowing the caller of the function to optionally change the password field in one go since an upsert without a password is valid, but an update may not want to update the password

**Comment by psibi** [comment]
Which version of persistent are you using ?

Note that the upsert behavior was fixed in version 2.7.0. You can see more details here: https://github.com/yesodweb/persistent/issues/613

**Comment by felixSchl** [comment]
Sorry, I should have checked that first! I started off with the yesod-postgres scaffold and it specified `persistent >= 2.0 && < 2.7`. `stack list-dependencies` reports version persistent-2.6.1 was installed. The new behavior looks sound to me, time to update :)
